cyral/rescource_cyral_listener.go

```diff
@@ -0,0 +1,177 @@
+package cyral
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/cyralinc/terraform-provider-cyral/client"
+)
+
+// Listener struct. Proto ref: https://github.com/cyralinc/wrapper-client-go/blob/main/messages/listeners.proto#L112
+// the XxxSettings and XxxListener fields are oneof in protocol buffer, I assume they should be individual attributes here.
+// For these, there should be one present in payload, if more one will be picked - so all marked as omitempty.
+// golang code will need to enforce this manually. Perhaps there is a assertOneOf function in the codebase?
+// I also assume it is good practise to keep sidecarId out of JSON since that is only used as a path parameter
+type SidecarListener struct {
+	SidecarId        string
+	ListenerId       string           `json:"id"`
+	UnixListener     UnixListener     `json:"unixListener,omitempty"`
+	TCPListener      TCPListener      `json:"tcpListener,omitempty"`
+	Multiplexed      bool             `json:"multiplexed,omitempty"`
+	MysqlSettings    MysqlSettings    `json:"mysqlSettings,omitempty"`
+	S3Settings       S3Settings       `json:"s3Settings,omitempty"`
+	DynamoDbSettings DynamoDbSettings `json:"dynamoDBSettings,omitempty"`
+}
+type UnixListener struct {
+	File string `json:"file"`
+}
+type TcpListener struct {
+	Host string `json:"host,omitempty"`
+	Port int    `json:"port"`
+}
+type MysqlSettings struct {
+	DbVersion    string `json:"dbVersion,omitempty"`
+	CharacterSet string `json:"characterSet,omitempty"`
+}
+type S3Settings struct {
+	ProxyMode bool `json:"proxyMode,omitempty"`
+}
+type DynamoDbSettings struct {
+	ProxyMode bool `json:"proxyMode,omitempty"`
+}
+
+// resourceSidecarListener returns the schema and methods for provisioning a sidecar listener
+// Sidecar listeners API is {{baseURL}}/sidecars/:sidecarID/listeners/:listenerID
+// GET {{baseURL}}/sidecars/:sidecarID/listeners/:listenerID (Get one listener)
+// GET {{baseURL}}/sidecars/:sidecarID/listeners/ (Get all listeners)
+// POST {{baseURL}}/sidecars/:sidecarID/listeners/ (Create a listener)
+// PUT {{baseURL}}/sidecars/:sidecarID/listeners/:listenerID (Update a listener)
+// DELETE {{baseURL}}/sidecars/:sidecarID/listeners/:listenerID (Delete a listener)
+func resourceSidecarListener() *schema.Resource {
+	return &schema.Resource{
+		Description:   "Manages [sidecar listeners](https://cyral.com/docs/sidecars/sidecar-listeners).",
+		CreateContext: resourceSidecarListenerCreate,
+		ReadContext:   resourceSidecarListenerRead,
+		UpdateContext: resourceSidecarListenerUpdate,
+		DeleteContext: resourceSidecarListenerDelete,
```

Comment:
I believe it's more "idiomatic" to use `CreateResource()`, `ReadResource()`, `UpdateResource()`, and `DeleteResoure()`, (see https://github.com/cyralinc/terraform-provider-cyral/blob/main/cyral/resource_cyral_repository_identity_map.go#L211) instead of creating ad-hoc functions.

This will also lead to simpler code, because you just need to implement `ReadFromSchema()` for create/update, and `WriteToSchema()` for read.

Reply:
Thanks @ricardorey10 for putting me on the right path, rewriting this.